### DISCUSSION TITLE
Send custom Header parameters using supabase-js

### DISCUSSION THREAD

**Discussion by WilliamFalci** [supabase/supabase#4755] [Questions] [answered]
Hi,

I will try to explain my case, I have to handle registered users and guest users, in case of the register user is pretty simple, we have already everything to handle it by auth functions (postgress ect) but in case of guest we need create all the structure.

So in my case I made various tables on postgress like guest, guest_settings ect, but when I have to made RLS policies I would like a custom header paraemeter where the value is the Guest ID and then retrieve it by a postgress function.

So here the "request" is possible in some way send a custom header parameter using supabase-js and retrieve it by postgress function to realize properly the various RLS policies in case of Guest users instead Registered User?

**Comment by mattWalters0** [comment]
I have a similar requirement to this.  I am using a shared link that contains a guid that allows anonymous users to access a page if the link has been shared with them.  I'd like to be able to implement the RLS to allow access if the guid is provided as a custom parameter.

**Comment by silentworks** [comment]
I can't see why you need a custom header to implement this. How is the guest user registered?

**Reply by WilliamFalci** [reply]
> I can't see why you need a custom header to implement this. How is the guest user registered?

This is the paradox, the guest can't be a registered user, this mean no reason to use supabase.auth.register /supabase.auth.login

To store guest data I use an HID (hardware identifier) stored in local storage / cookie, with an function of re-generation and indetifier if user change some hardware component.

So the guest is stored in a simple table with an unique id and hid, from where I store / update and compare the HID generated from frontend to indentify the guest.

This work at 100% but is a "limit" for the RLS, this because without send the Guest ID I can't set properly the rules. So in this case Im forced to register the guest with a fake registration, where for example I register it by email registration and the email is HID@hid.it and then I have to fix all the logic for work in this way. But can u imagine how many user the db will have?

So from my point of view the right handling would be:

1. Frontend Generate the HID
2. Frontend check if user have already an HID stored in localstorage / cookies
3. If user haven't any HID stored in localstorage / cookies mean is a new guest, then save the new HID to the table and init guest data
4. If user have already an HID stored in localstorage / cookies mean was already an guest, get Guest data using Stored HID, if Generate HID is different then Stored HID, then update the HID on DB and localstorage / cookies
5. Use Guest ID and User Type into RLS policy to allow the SELECT and UPDATE only for the proper data.

But I can't post Guest ID and User Type in a custom variable on headers, then I can't set properly the RLS policies.

**Comment by XHMM** [comment]
Have the same requirement too.  

I found we can [get http headers](https://postgrest.org/en/stable/api.html#accessing-request-path-and-method) in postgrest doc, and I tried create policy like this,  but seems not work (new to postgresql) :

```sql
CREATE POLICY "test" ON public.forms FOR UPDATE 
USING ((((current_setting('request.headers'::text, true))::json ->> email) = email)) 
WITH CHECK ((((current_setting('request.headers'::text, true))::json ->> email) = email));
```

**Reply by WilliamFalci** [reply]
> Have the same requirement too.
> 
> I found we can [get http headers](https://postgrest.org/en/stable/api.html#accessing-request-path-and-method) in postgrest doc, and I tried create policy like this, but seems not work (new to postgresql) :
> 
> ```sql
> CREATE POLICY "test" ON public.forms FOR UPDATE 
> USING ((((current_setting('request.headers'::text, true))::json ->> email) = email)) 
> WITH CHECK ((((current_setting('request.headers'::text, true))::json ->> email) = email));
> ```

In case of user registered on supabase.auth you can simple use the function auth.email() in your RLS policy

In my scenario the user isn't registered into supabase.auth so I need a way to post an custom parameter with the value I need and then get it from an function inside the RLS policy (like what you did)

**Comment by GaryAustin1** [comment, marked as answer]
When you init the Supabase client you can pass in a header with the headers parameter.
https://supabase.com/docs/reference/javascript/initializing#with-additional-parameters
`headers: { 'myheader': 'extrausernumber' },`
This is supposed to be sent out on all calls and was fixed to do so a month or so ago I believe.  https://github.com/supabase/supabase-js/pull/311

In your RLS policy you can use this as documented in POSTGREST: https://postgrest.org/en/stable/api.html#http-context
`SELECT current_setting('request.headers', true)::json->>'myheader'`
or put it in a function and use it like the auth.uid() function

Note: I've not tested any of this.
Edit:  Also note that your plan is not secure so you should not have privileged info behind this method.

**Reply by WilliamFalci** [reply]
What about if supabase is init before we get the values to parse into headers?

What do u want mean isn't secure? Is Like the auth.uid() concept what change Is just the value

**Reply by GaryAustin1** [reply]
I believe you can have two clients if you can't get your header info generated at the start (like when dealing with different schemas).  You might be able to just rerun createClient if you have not signed in, I know it does not error if called twice before sign in.

You are not secure because you have a plain text code being stored in localstorage and sent out on the network, also viewable in browser console.  Auth uses JWT tokens that have to be refreshed and get regenerated on sign in.  There is no information on the client or network calls that could be lifted and used to make fake requests later.

Edit: On secure, the issue is really how secure do you need it to be.  It will certainly protect from random use of the device ID you are creating.  No one is likely to guess your device IDs if they are complex enough.  So a bad actor would have to get it from the computer or wire and then use it.  If you are just using for guest settings, counts, etc. then probably no big deal.

**Reply by WilliamFalci** [reply]
> I believe you can have two clients if you can't get your header info generated at the start (like when dealing with different schemas). You might be able to just rerun createClient if you have not signed in, I know it does not error if called twice before sign in.
> 
> You are not secure because you have a plain text code being stored in localstorage and sent out on the network, also viewable in browser console. Auth uses JWT tokens that have to be refreshed and get regenerated on sign in. There is no information on the client or network calls that could be lifted and used to make fake requests later.
> 
> Edit: On secure, the issue is really how secure do you need it to be. It will certainly protect from random use of the device ID you are creating. No one is likely to guess your device IDs if they are complex enough. So a bad actor would have to get it from the computer or wire and then use it. If you are just using for guest settings, counts, etc. then probably no big deal.

I tested it and work but I got an issue, I reported here:

[https://github.com/supabase/supabase/issues/4797](https://github.com/supabase/supabase/issues/4797)

**Reply by GaryAustin1** [reply]
Edit: I think your problem is you are checking your passed header against both the new value coming in for that column and the old column and they are different as you are changing it. You will need either do a new insert with the new hid, or pass the old header value for the update and have the update RLS  do “using” with the existing hid and have a “with check” as just true as you are changing it.

**Reply by soedirgo** [reply]
Marking this one as answered as the custom headers part is working.
